readme.txt
This is version 0.29 of Daily Cartoons; please check iandouglas.com for newer
versions.

What is this Plugin?
- This plugin will let you view your favorite cartoons from around the
  Internet based on a little configuration effort on your part. Other plugins
  just show you the daily Dilbert cartoon, and this script extends that
  functionality to view multiple cartoons on a single page.


Contact:
- If you have questions, problems, suggestions, or feedback, or want me to
  figure out how to add your favorite cartoon, email ian.douglas@iandouglas.com
  or wild98@gmail.com and I'll do whatever I can if you ask nicely.


Installation:
- Unpack the archive into your e107_plugins folder; this should create a
  folder in your plugins folder called "dailycartoons" with other subfolders
  called 'cache' and '404comics'.
- I've included a sample cartoons.php configuration with 22 of my favorite
  cartoons, which admittedly may take a minute or two for you to load the
  first time, and assumes you want to cache the images.
- Test it by going to http://yoursite/e107_plugins/dailycartoons/
- If that works, open your Admin panel, click on the 'Site Links' icon, then
  'Add a new Link' and enter the full URL you just tested
- Now you should be able to add the link into your menus so your users can
  click a link instead of bookmarking a plugin page.
- By default, the .htaccess that gets built for e107 blocks search engines
  from indexing any pages under the 'e107_plugins' path


New in Version ____
- please consult the history.txt file for new information on changes in this
	and other versions.


Adding New Cartoons:
- Modify the cartoons.php file, and add a new element to the big array. The
  pieces needed for each cartoon include:
  - 'name' - give it any name you like; this will show up on the web page
  - 'site' - this is the URL where you'd visit to view the cartoon before 
    this plugin came along
  - 'url' - this is what the URL's look like when browsing their archive. Use
    the {{{DATE}}} element to insert the date.
  - 'dateformat' - the format of the date they use for browsing archived
    cartoons. For example, Dilbert uses a format of "Ymd" which is a YYYYMMDD
    format. Foxtrot uses "Y/m/d" which is a YYYY/MM/DD format. This is how the
    script will change the {{{DATE}}} element in 'url' for looking at archived
    pages.
  - 'regexp' - this is a regular expression for how to actually get the name
    of the cartoon graphic from the originating server so we can view it with
    the plugin. Teaching regular expressions is way outside the scope of this
    document.
  - 'cartoon' - this is how the script incorporates the cartoon image into the
    plugin page that gets drawn. The markers like {{{1}}} or {{{2}}} mark
    where we'll fill in the pattern(s) we found with the regular expression.
  - 'cache' - set to 1 if you want to save a copy of the .gif or .jpg cartoon
    in the 'cache' folder or not. If your site gets a lot of traffic, do us
    all a favor and turn this option on (1). Setting it to 0 will turn caching
    off.
The following setting is optional and can be excluded if you don't want to use
it:
  - '404' - this is the name of the graphic in the '404comics' folder that you
    want to show users when a regular expression can't find a match or the
    cartoon archive page is missing (ie: you don't have a cartoon cached and
    the original archive page doesn't go back that far). The string value here
    is case-sensitive if you're hosted on a Unix/Linux platform. I've included
    a 404 comic for each of the cartoons in my example file.
    Note that if this option is excluded, your users will see text strings
    explaining why a comic could not be viewed.
    
If you can't figure out the regular expression for a particular comic, contact
me and I'll help out as best I can. Don't send me messages saying "Can you add
______?" because that's just rude. Help me out a little and at least provide a
URL to where the cartoons come from and how to access an archive of previous
cartoons. I've included a LOT of cartoons as of v0.20 which are some of my
all-time favorites like Adam@Home, Dilbert, Foxtrot, Drabble, Rose is Rose,
and so on. Most of these, you'll see from my example cartoons.php, either come
from comics.com or gocomics.com, so you can just copy and paste my examples
for other comics from these services that you'd like to see. If you want to
add comics from a totally different comic source, I can help if you ask
nicely.


Limitations:
- I added a 10-minute timeout in the script using ini_set() in case you load
  up dozens and dozens of cartoons and your server downloads them for the
  first time for a visitor, since downloading the pages then extracting the
  image and saving it on your system can be time-consuming if your server or
  bandwidth is slow. My workaround is to find my lowest point of traffic
  overnight, and have a cron job running that calls 'wget' to fetch my
  /dailycartoon/ page so the graphics get cached.
- This plugin cannot currently get into site that require a login of any type.
  That's not even on the to-do list.


To-Do List:
- make the configuration a little less daunting

Credit where it's due:
- This e107 plugin was developed by modifying Richard Perry's Daily Dilbert
  Plus plugin and expanding it to check multiple cartoon sources. Thanks to
  Richard for making the plugin freely available to modify like this. You can
  contact him at http://www.greycube.com
- thanks to the many comics services out there that don't have hot-linking
  prevention turned on to allow geeks like me to develop something like this.



history.txt
New in Version 0.29
- fopen() workaround
	Version 0.29 uses fsockopen() for downloading comic pages and cartoon
	images. The code will autodetect whether "allow_url_fopen" is turned on in
	php.ini or not, and if not, will fallback to use the fsockcomm.
	If this autodetection does not work for you, look for this line of code
	around line 78:
		//$fopen_check = 'Off' ;
	and remove the // comment markers. This will force the code to use the
	fsockopen() workaround.

New in Version 0.25
- cache checking
	Version 0.20 wouldn't check to see if it had already cached the cartoon
	before opening the cartoon's web page to extract the filename to see if it
	already existed. Adding code to check the cache folder *before* opening the
	cartoon's web page has significantly decreased the processessing time.
	The limitation on this cache check is that if you change the 'name' element
	for a specific cartoon, is that cached items use the 'name' element as part
	of the cache filename. If you alter the 'name' element, the code will fail
	to locate older cached files. On operating systems that have case-sensitive
	file systems, this can impact even changing a cartoon name from "My Cartoon"
	to "my cartoon". Windows-hosted systems may or may not be affected by
	case-sensitivity issues, but Unix/Linux/BSD systems generally will be
	impacted.
- removed "PC and Pixel" cartoon from cartoons.php
	The author has slipped behind a few days and was always generating a 404
	error when trying to fetch a cartoon based on "today's" date.

New in Version 0.20
- caching
  The tarball now includes a 'cache' folder which you'll need to set the
  permissions to 777 using the CHMOD command in your FTP/ssh program. The
  script will *attempt* to save a copy of the cartoon graphic on your server,
  but after a while this could really add up to a lot of disk space, so be
  sure your hosting account can handle the load, or just delete older graphics
  every month or so.
- handling 404 and regular expression issues
  If the script can't apply a regular expression on a downloaded cartoon page,
  or if the cartoon page is simply unable to be opened, it will show a graphic
  of your choice from the '404comics' page. I used a graphic editor with the
  'Comic Sans MS' font to create my own goofy 'comic is missing' images.




index.php
<?php
/*
	e107 website system, daily comics
	based heavily on Daily Dilbert Plus by Richard Perry, http://www.greycube.com/
	Richard gave credit to Kimmo of http://www.jernstrom.org/

	v0.29, September 23, 2006
	- workaround for those that don't have fopen() support for opening URL's

	v0.25, September 15, 2006
	- check cache for a cartoon before loading cartoon's web page
	- changes: check cache for cartoon file existance BEFORE downloading the cartoon URL
		- this should speed up processing *significantly*
	- include the regexp/404 text message under the 404 error graphic so the user can see why the cartoon didn't load

	v0.20, August 24, 2006
	- changes: added cache support

	v0.10, August 23, 2006
	- initial release
*/
require_once("../../class2.php") ;
require_once(HEADERF) ;
ini_set('max_execution_time', '600') ;
$version = "0.29" ;
$index = 0 ;
$CARTOONS = Array() ;
include("cartoons.php") ;
// attempt to make the cache folder and chmod 777 if it doesn't already exist
// we'll use the @ prefix to supress any errors
@mkdir("cache") ;
@chmod(0777,"cache") ;
//[[[ date processing
$today = date("Y-m-d") ;
$todaySec = date("U") ;
$date = $HTTP_GET_VARS['date'] ;
if (!$date) {
	$date = date("Y-m-d") ;
	$prevday = date("Y-m-d",strtotime("yesterday")) ;
	$nextday = date("Y-m-d",strtotime("tomorrow")) ;
}
foreach ($CARTOONS as $index => $cartoon) {
	$CARTOONS[$index]['date'] = date($cartoon['dateformat'],strtotime($date)) ;
	//	$CARTOONS[$index]['prevday'] = date($cartoon['dateformat'],strtotime($prevday)) ;
	//	$CARTOONS[$index]['nextday'] = date($cartoon['dateformat'],strtotime($nextday)) ;
}

$prevday = date("Y-m-d",mktime (0,0,0,substr($date, 5, 2),substr($date, 8, 2)-1,substr($date, 0, 4))) ;
$nextday = date("Y-m-d",mktime (0,0,0,substr($date, 5, 2),substr($date, 8, 2)+1,substr($date, 0, 4))) ;
$nextdaySec = date("U",mktime (0,0,0,substr($date, 5, 2),substr($date, 8, 2)+1,substr($date, 0, 4))) ;
//]]]

$text = '<div style="font-family: tahoma; font-size: 9px; text-align:center">Daily Cartoons v'.$version.' by <a href="http://iandouglas.com/">Ian Douglas</a><br />Based on Daily Dilbert Plus By Richard Perry<br /></div><table style="width:95%; border:0;"><tr><td style="text-align:left"><b><a href="'.$PHP_SELF.'?date='.$prevday.'">Previous Day</a></b></td><td style="text-align:right">'.($nextdaySec <= $todaySec ? '<b><a href="'.$PHP_SELF.'?date='.$nextday.'">Next Day</a></b>' : '').'</td></tr></table><br />' ;

$template = '<table style="width:95%; border:0;"><tr><td style="text-align:center"><a href="{{{SITE}}}" target="new">{{{CARTOON}}}</a></td></tr></table><br />' ;
foreach ($CARTOONS as $index => $cartoon) {
	$text .= "<p>{$cartoon['name']}<br>\n" ;
	$done = 0 ;
	$cartoon_text = $template ;
	if ($cartoon['cache']) {
		$dh = opendir("cache") ;
		$cache_match = "{$cartoon['name']}__".$date."__" ;
		while ($file = readdir($dh)) {
			if (strstr($file,$cache_match) !== false) {
				// build cartoon text here
				$cartoon_file = "cache/".$file ;
				$cartoon_text = str_replace("{{{CARTOON}}}",$cartoon['cartoon'],$cartoon_text) ;
				$cartoon_text = str_replace("{{{FETCH}}}",$cartoon_file,$cartoon_text) ;
				$done = 1 ;
				break ;
			}
		}
		closedir($dh) ;
	}
	if (!$done) {
		$url = str_replace("{{{DATE}}}",$cartoon['date'],$cartoon['url']) ;
		$fopen_check = ini_get('allow_url_fopen') ;
		//$fopen_check = 'Off' ;
		if ($fopen_check == 'On' || $fopen_check == 1 || $fopen_check == 'true') {
			$fh = @fopen($url,"r") ;
		} else {
			preg_match('/http:\/\/([^\/]*)(.*)/i', $url, $matches) ;
			$fh = fsockopen($matches[1], 80, $errno, $errstr, 10) ;
			fputs($fh, "GET /{$matches[2]} HTTP/1.0\r\n") ;
			fputs($fh, "Host: {$matches[1]}\r\n") ;
			fputs($fh, "Referer: http://{$matches[1]}\r\n") ;
			fputs($fh, "User-Agent: DailyCartoons v$version\r\n\r\n");
		}

		if ($fh) { //[[[
			while (!feof($fh)) {
				$tmp .= fread($fh, 2048) ;
			}
			fclose ($fh) ;
			$bytes = strlen($tmp) ;
			// draw image here
			$tmp = str_replace("\r"," ",$tmp) ;
			$tmp = str_replace("\n"," ",$tmp) ;
			preg_match($cartoon['regexp'],$tmp,$matches) ;
			if ($matches[0]) { //[[[ regexp worked
				$cartoon_file = $cartoon['fetch'] ;
				$matchindex = 0 ;
				for($i=1; $i < count($matches); $i++) {
					$cartoon_file = str_replace("{{{".$i."}}}",$matches[$i],$cartoon_file) ;
				}
				if ($cartoon['cache']) {//[[[
					$bits = explode("/",$cartoon_file) ;
					$cartoon_cachefile = $bits[count($bits)-1] ; // get just the last filename portion
					$cartoon_cachefile = str_replace(".","_",$cartoon['name']) . "__" . $date . "__" . $cartoon_cachefile ;
					$cartoon_cachefile = str_replace(" ","_",$cartoon_cachefile) ;
					$cartoon_cachefile = str_replace("\\","",$cartoon_cachefile) ;
					$cartoon_cachefile = str_replace(":","",$cartoon_cachefile) ;
					$cartoon_cachefile = str_replace("/","",$cartoon_cachefile) ;
					// this will create a filename like "Daily_Dilbert__2006-08-24__dilbert20060824.gif"
					if (file_exists("cache") && file_exists("cache/".$cartoon_cachefile)) {
						$cartoon_file = "cache/".$cartoon_cachefile ;
					} elseif (file_exists("cache")) {
						// save a copy of the cartoon
						if ($fopen_check == 'On' || $fopen_check == 1 || $fopen_check == 'true') {
							$fh_cartoon = fopen($cartoon_file,"rb") ;
						} else {
							preg_match('/http:\/\/([^\/]*)/(.*)/i', $cartoon_file, $matches2) ;
							$fh_cartoon = fsockopen($matches2[1], 80, $errno, $errstr, 10) ;
							if (substr($matches2[1],-1,1) == '/') {
								$matches2[1] = substr($matches2[1],0,strlen($matches2[1] - 1)) ;
							}
							fputs($fh_cartoon, "GET /{$matches2[2]} HTTP/1.0\r\n") ;
							fputs($fh_cartoon, "Host: {$matches2[1]}\r\n") ;
							fputs($fh_cartoon, "Referer: http://{$matches2[1]}\r\n") ;
							fputs($fh_cartoon, "User-Agent: DailyCartoons v$version\r\n\r\n");
						}
						if ($fh_cartoon) {
							$cartoon_binary = '' ;
							$byte_count = 0 ;
							while (!feof($fh_cartoon)) {
								$cartoon_binary .= @fread($fh_cartoon,1024) ;
								$byte_count++ ;
							}
							fclose($fh_cartoon) ;
							if ($fopen_check != 'On' && $fopen_check != 1 && $fopen_check != 'true') {
								// strip headers from cached graphic
								list($junk,$cartoon_binary) = explode("\r\n\r\n",$cartoon_binary) ;
							}
							$fh_cartoon = @fopen("cache/".$cartoon_cachefile,"wb") ;
							if ($fh_cartoon) {
								fwrite($fh_cartoon,$cartoon_binary,$byte_count) ;
								fclose($fh_cartoon) ;
							}
							$cartoon_file = "cache/".$cartoon_cachefile ;
						}
					}
				}//]]]
				$cartoon_text = str_replace("{{{CARTOON}}}",$cartoon['cartoon'],$cartoon_text) ;
				$cartoon_text = str_replace("{{{FETCH}}}",$cartoon_file,$cartoon_text) ;
				// reapply the {{{#}}} filters, if any
				for($i=1; $i < count($matches); $i++) {
					$cartoon_text = str_replace("{{{".$i."}}}",$matches[$i],$cartoon_text) ;
				}
				//]]]
			} else {//[[[
				// preg_match didn't find any matches
				$cartoon_text = str_replace("{{{CARTOON}}}",(isset($cartoon['404']) ? '<img src="404comics/'.$cartoon['404'].'" border="0" alt="comic not found (regexp failed)" /><br />The regular expression for this cartoon did not find a match.' : "The regular expression for this cartoon did not find a match."),$cartoon_text) ;
			}//]]]
			//]]]
		} else { //[[[
			// couldn't open page, either fopen isn't allowed to open url's or the page is 404, etc
			$cartoon_text = str_replace("{{{CARTOON}}}",(isset($cartoon['404']) ? '<img src="404comics/'.$cartoon['404'].'" border="0" alt="comic not found (404)" /><br />Could not open the cartoon web page (404 not found)' : "Could not open <a href=\"$url\">$url</a>, 404 Not found."),$cartoon_text) ;
		}//]]]
	}
	$cartoon_text = str_replace("{{{SITE}}}",$cartoon['site'],$cartoon_text) ;
	$text .= "\n$cartoon_text</p>\n" ;
}
$text .= '<table style="width:95%; border:0;"><tr><td style="text-align:left"><b><a href="'.$PHP_SELF.'?date='.$prevday.'">Previous Day</a></b></td><td style="text-align:right">'.($nextdaySec <= $todaySec ? '<b><a href="'.$PHP_SELF.'?date='.$nextday.'">Next Day</a></b>' : '').'</td></tr></table><br />' ;

$ns->tablerender("Daily Cartoons", $text);

echo "\r\n<!--\r\n = Daily Cartoons for E107 v1.0, by Ian Douglas, www.iandouglas.com, based on DAILY DILBERT PLUS BY RICHARD PERRY =\r\n -->\r\n";
require_once(FOOTERF);
?>
